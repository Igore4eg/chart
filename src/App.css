.App {
  margin:  auto 50px;
}
.charts {
  
  display: flex;
  flex-direction: column;
  row-gap: 0.5em;
}
.wraperString {
  display: flex;
  justify-content: end;
  align-items: center;
  column-gap: 20px;
}
.lineName {

}
.wrapperLine {
  width: 65%;
  background-color: rgb(226, 226, 226);
}
.line {
  text-align: center;
  background-color:rgb(79, 111, 148)
}