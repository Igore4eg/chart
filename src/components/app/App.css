.App {
  margin:  auto 50px;
  font-size: 20px;
}
.charts {
  max-width: 1100px;
  display: flex;
  flex-direction: column;
  row-gap: 0.5em;
}
