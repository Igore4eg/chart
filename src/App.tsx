import { useState, useEffect } from "react";

import './App.css';

interface IInitialState {
  name: string;
  time: number;
}

const initialState: IInitialState[]  = [
  { name: 'Landing Page', time: 7.4 },
  { name: 'Configurator', time: 0.2 },
  { name: 'Check-out', time: 7.0 },
  { name: 'Deal', time: 3.8 },
  ];

function App() {
  const [chart, setChart] = useState(initialState);

  useEffect(() => {
    const timerId = setInterval(updateChart, 30000);

    return () => {
      clearInterval(timerId)
    }
  }, [])
  
  const allTime: number = +chart.reduce((prev, cur) => {
    return prev + cur.time
  }, 0).toFixed(2);

  const updateChart = () => {
    const newChart: IInitialState[] = chart.map(value => {
      let newTime: number = +(Math.random() * (10 - 0.1) + 0.1).toFixed(1)
      return {name: value.name, time: newTime}
    })
    setChart(newChart)
  }

  let prevW: number = 0;
  const charts = chart.map((value, index) => {
      const indent = prevW;
      const w = value.time/allTime*100;
      prevW += w;
      return (
        <View key={value.name} 
          name={value.name}
          time={value.time} 
          width={w} 
          ml={indent}/>
      )
    })

  return (
    <div className="App">
      <p>SPENT TIME (SECONDS)</p>
      <div className="charts">
        {charts}
      </div>
      <button onClick={updateChart}>RND</button>
    </div>
  );
}

type ViewProps = {
  name: string;
  time: number;
  width: number;
  ml: number;
}

const View = ({name, time, width, ml}: ViewProps,  ) => {
  
  return(
    <div className="wraperString">
      <div className="lineName">{name}</div>
      <div className="wrapperLine">
        <div className="line" style={{width: `${width}%`, marginLeft: `${ml}%`}}>{`${time}`}</div>
      </div>
    </div>
  )
}

export default App;
