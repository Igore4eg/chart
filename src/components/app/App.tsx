import { useState, useEffect } from "react";
import ViewChart from '../viewChart/ViewChart'

import './App.css';

interface IInitialState {
	name: string;
	time: number;
};

const initialState: IInitialState[]  = [
	{ name: 'Landing Page', time: 7.4 },
	{ name: 'Configurator', time: 0.2 },
	{ name: 'Check-out', time: 7.0 },
	{ name: 'Deal', time: 3.8 },
];

function App() {

	const [chart, setChart] = useState(initialState);

	useEffect(() => {
		const timerId = setInterval(updateChart, 30000);
		
		return () => {
		clearInterval(timerId)
		}
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	const updateChart = () => {
		const newChart: IInitialState[] = chart.map(value => {
			let newTime: number = +(Math.random() * (10 - 0.1) + 0.1).toFixed(1)
			return {name: value.name, time: newTime}
		})
		setChart(newChart)
	};

	const renderCharts = ()=> {
		let prevW: number = 0;
		const allTime: number = +chart.reduce((prev, cur) => {
		return prev + cur.time
		}, 0).toFixed(2);
		const charts = chart.map((value) => {
			const indent = prevW;
			const w = value.time/allTime*100;
			prevW += w;
			return (
				<ViewChart key={value.name} 
					name={value.name}
					time={value.time} 
					width={w} 
					ml={indent}/>
			)
		});
		return charts;
	};

	const allChart = renderCharts();
	
	return (
		<div className="App">
		<p>SPENT TIME (SECONDS)</p>
		<div className="charts">
			{allChart}
		</div>
		<button data-testid="btn-rnd" onClick={updateChart}>RND</button>
		</div>
	);
}

export default App;