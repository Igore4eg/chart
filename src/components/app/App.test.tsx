import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';

test('renders landing page', () => {
  const {container} = render(<App />);
  const chartElement = screen.getByText(/landing page/i);
  expect(chartElement).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('render after RND button', () => { 
    const {container} = render(<App />);
    const btn = screen.getByTestId('btn-rnd');
    fireEvent.click(btn);
    expect(container.querySelectorAll('.wraperString').length).toBe(4);
});