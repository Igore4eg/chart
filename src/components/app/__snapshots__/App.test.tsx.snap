// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders landing page 1`] = `
<div>
  <div
    class="App"
  >
    <p>
      SPENT TIME (SECONDS)
    </p>
    <div
      class="charts"
    >
      <div
        class="wraperString"
      >
        <div
          class="lineName"
        >
          Landing Page
        </div>
        <div
          class="wrapperLine"
        >
          <div
            class="line"
            style="width: 40.217391304347835%; margin-left: 0%;"
          >
            7.4
          </div>
        </div>
      </div>
      <div
        class="wraperString"
      >
        <div
          class="lineName"
        >
          Configurator
        </div>
        <div
          class="wrapperLine"
        >
          <div
            class="line"
            style="width: 1.0869565217391306%; margin-left: 40.217391304347835%;"
          >
            0.2
          </div>
        </div>
      </div>
      <div
        class="wraperString"
      >
        <div
          class="lineName"
        >
          Check-out
        </div>
        <div
          class="wrapperLine"
        >
          <div
            class="line"
            style="width: 38.04347826086957%; margin-left: 41.30434782608697%;"
          >
            7
          </div>
        </div>
      </div>
      <div
        class="wraperString"
      >
        <div
          class="lineName"
        >
          Deal
        </div>
        <div
          class="wrapperLine"
        >
          <div
            class="line"
            style="width: 20.652173913043477%; margin-left: 79.34782608695653%;"
          >
            3.8
          </div>
        </div>
      </div>
    </div>
    <button
      data-testid="btn-rnd"
    >
      RND
    </button>
  </div>
</div>
`;
